### ISSUE TITLE
LoweringError when using sqrt in CUDA

### ISSUE THREAD

**Issue by mckwxp** [numba/numba#4000]
## Reporting a bug

- [x] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

I am using python 3.6.8, numba 0.43.1, and cudatoolkit 9.0. The GPU I'm using is NVIDIA GeForce 1050 Ti. When I run the following code I got an error.

```python
from numba import cuda
from math import sqrt
import numpy as np

N = 1000

@cuda.jit
def foo(x, y):
    i = cuda.grid(1)
    if i < N:
        y[i] = sqrt(x[i])

a = np.arange(N)
b = np.zeros(N)
foo[32,32](a, b)
```
The error message is:
```
Traceback (most recent call last):
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/errors.py", line 627, in new_error_context
    yield
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/lowering.py", line 258, in lower_block
    self.lower_inst(inst)
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/lowering.py", line 301, in lower_inst
    val = self.lower_assign(ty, inst)
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/lowering.py", line 454, in lower_assign
    return self.lower_expr(ty, value)
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/lowering.py", line 913, in lower_expr
    res = self.lower_call(resty, expr)
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/lowering.py", line 706, in lower_call
    res = self._lower_call_normal(fnty, expr, signature)
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/lowering.py", line 877, in _lower_call_normal
    impl = self.context.get_function(fnty, signature)
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/targets/base.py", line 551, in get_function
    return self.get_function(fn, sig, _firstcall=False)
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/targets/base.py", line 553, in get_function
    raise NotImplementedError("No definition for lowering %s%s" % (key, sig))
NotImplementedError: No definition for lowering <built-in function sqrt>(int64,) -> float64

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test.py", line 15, in <module>
    foo[32,32](a, b)
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/cuda/compiler.py", line 765, in __call__
    kernel = self.specialize(*args)
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/cuda/compiler.py", line 776, in specialize
    kernel = self.compile(argtypes)
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/cuda/compiler.py", line 792, in compile
    **self.targetoptions)
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/compiler_lock.py", line 32, in _acquire_compile_lock
    return func(*args, **kwargs)
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/cuda/compiler.py", line 62, in compile_kernel
    cres = compile_cuda(pyfunc, types.void, args, debug=debug, inline=inline)
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/compiler_lock.py", line 32, in _acquire_compile_lock
    return func(*args, **kwargs)
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/cuda/compiler.py", line 51, in compile_cuda
    locals={})
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/compiler.py", line 941, in compile_extra
    return pipeline.compile_extra(func)
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/compiler.py", line 372, in compile_extra
    return self._compile_bytecode()
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/compiler.py", line 872, in _compile_bytecode
    return self._compile_core()
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/compiler.py", line 859, in _compile_core
    res = pm.run(self.status)
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/compiler_lock.py", line 32, in _acquire_compile_lock
    return func(*args, **kwargs)
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/compiler.py", line 253, in run
    raise patched_exception
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/compiler.py", line 244, in run
    stage()
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/compiler.py", line 731, in stage_nopython_backend
    self._backend(lowerfn, objectmode=False)
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/compiler.py", line 681, in _backend
    lowered = lowerfn()
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/compiler.py", line 668, in backend_nopython_mode
    self.metadata)
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/compiler.py", line 1063, in native_lowering_stage
    lower.lower()
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/lowering.py", line 177, in lower
    self.lower_normal_function(self.fndesc)
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/lowering.py", line 218, in lower_normal_function
    entry_block_tail = self.lower_function_body()
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/lowering.py", line 243, in lower_function_body
    self.lower_block(block)
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/lowering.py", line 258, in lower_block
    self.lower_inst(inst)
  File "/home/marco/miniconda3/lib/python3.6/contextlib.py", line 99, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/errors.py", line 635, in new_error_context
    six.reraise(type(newerr), newerr, tb)
  File "/home/marco/miniconda3/lib/python3.6/site-packages/numba/six.py", line 659, in reraise
    raise value
numba.errors.LoweringError: Failed in nopython mode pipeline (step: nopython mode backend)
No definition for lowering <built-in function sqrt>(int64,) -> float64

File "test.py", line 11:
def foo(x, y):
    <source elided>
    if i < N:
        y[i] = sqrt(x[i])
  ^

[1] During: lowering "$18.5 = call $18.1($18.4, func=$18.1, args=[Var($18.4, test.py (11))], kws=(), vararg=None)" at test.py (11)
```

**Comment by mckwxp** [comment]
Just to add, this happens for many functions from the math module listed on https://numba.pydata.org/numba-doc/dev/cuda/cudapysupported.html#math

**Comment by stuartarchibald** [comment]
Thanks for the report. The cause of the problem you are seeing is that the CUDA math functions that are bonded to the `math` module (for ease) only support operating on floating point types and there is no automatic type cast from e.g. integer input to some floating point input. In the sample provided, `a` is defined through `np.arange` which causes it to be an integer type, and it is this on which the `sqrt` call is made. Changing the code to:
```python
from numba import cuda
from math import sqrt
import numpy as np

N = 1000

@cuda.jit
def foo(x, y):
    i = cuda.grid(1)
    if i < N:
        y[i] = sqrt(x[i])

a = np.arange(float(N))
b = np.zeros(N)
foo[32,32](a, b)
```
works fine.

I am however going to mark this as a bug because this is an easily detected set of circumstances and the error message should be improved.

**Comment by mckwxp** [comment]
I see. Thanks for the reply.

**Comment by gmarkall** [comment]
This will be fixed by #6152. Closing this issue as it's duplicated by #4618.
